### PR TITLE
7068 Remove the ability to set the user's language via the API

### PR DESCRIPTION
**Pull request by jkuester**
# Description

Removes support for adding/modifying a user's language setting via the `/api/v1/users` endpoint.  Also updates the login controller to not modify the user's language setting with the locale provided on the request. 

The users service will no longer support adding/modifying/retrieving the `language`/`language.code` settings for a user.

Updates for the endpoint documentation are included in: https://github.com/medic/cht-docs/pull/485

One question I do have is if we need to include an api migration script with these changes to remove the language property from the existing settings for each user in the medic db?  I do not think this data is getting used anymore, but I am not sure if this is the kind of thing that usually gets a migration script...

Closes medic/cht-core#7068

# Code review checklist
- [x] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [x] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [x] Tested: Unit and/or e2e where appropriate
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: api/src/controllers/login.js [hunk 1]
```diff
@@ -182,18 +175,16 @@ const setCookies = (req, res, sessionRes) => {
         .getUserSettings(userCtx)
         .catch(err => {
           if (err.status === 404 && auth.isDbAdmin(userCtx)) {
-            return users.createAdmin(userCtx).then(() => auth.getUserSettings(userCtx));
+            return users.createAdmin(userCtx);
```
Comment (jkuester) [inline api/src/controllers/login.js:178]: I is possible that this call to `getUserSettings` on line 175 is unnecessary now (since the actually settings do not get used here).  However, I have left it in since the 404 for that call is what triggers the `createAdmin` logic. If there is a better way to check that a DbAdmin needs to be initialized, we could do that here instead.

Comment (jkuester) [reply api/src/controllers/login.js:178]: Either way, it seemed that the second call to getUserSettings (to retrieve the newly created admin's settings) was useless now, so I removed it.

File: api/src/controllers/login.js [hunk 2]
```diff
@@ -274,7 +265,7 @@ const loginByToken = (req, res) => {
       }
 
       return tokenLogin.resetPassword(userId).then(({ user, password }) => {
-        req.body = { user, password };
+        req.body = { user, password, locale: req.body.locale };
```
Comment (jkuester) [inline api/src/controllers/login.js:268]: This may have been a latent bug.  Several of the login tests started failing after I made the above changes and the ones involving token login (e.g. `should login the user when token is valid`) were not fixed even after I added the locale to the request body in the tests.  After some debugging I determined that the locale was getting dropped from the request body here at this line before `setCookies` is called.

File: api/tests/mocha/controllers/login.spec.js [hunk 3]
```diff
@@ -699,8 +697,8 @@ describe('login controller', () => {
         chai.expect(auth.isDbAdmin.args[0]).to.deep.equal([userCtx]);
         chai.expect(users.createAdmin.callCount).to.equal(1);
         chai.expect(users.createAdmin.args[0]).to.deep.equal([userCtx]);
-        chai.expect(auth.getUserSettings.callCount).to.equal(2);
-        chai.expect(auth.getUserSettings.args).to.deep.equal([[userCtx], [userCtx]]);
+        chai.expect(auth.getUserSettings.callCount).to.equal(1);
+        chai.expect(auth.getUserSettings.args).to.deep.equal([[userCtx]]);
```
Comment (jkuester) [inline api/tests/mocha/controllers/login.spec.js:701]: These assertions changed because I removed the second call to getUserSettings for retrieving the newly created admin's settings.

**Comment by garethbowen** [comment]
@dianabarsan Could you please review this when you get back? I'm slightly concerned now that there's no easy way for someone to choose a language when using token login. The locale selector is there, but the script attempts to log you in and redirect you to the app before you've had a chance to click it. This PR potentially makes this situation worse, because now an admin can't set it in the DB. I think this is still the right approach and maybe we have to look at a separate issue for changing the way token login works so the user has time to select a language and click a button to go to the app.

File: api/src/services/users.js [hunk 4]
```diff
@@ -34,7 +34,6 @@ const RESTRICTED_SETTINGS_EDITABLE_FIELDS = [
   'fullname',
   'email',
   'phone',
-  'language',
   'known',
```
Comment (dianabarsan) [inline api/src/services/users.js:37]: Unfortunately, this change breaks backwards compatibility. 

It is causing an e2e test in this build to fail, which involves changing the user's language locally. 
Webapp will try to call this endpoint to change the language on the server, and actually receive a 400, because no other changes are sent except language. 
The same happens when trying to update a user in the admin app. 
For offline users, it's worse as we filter which fields are editable and which are not, so a request that contains the `language` field will receive a 401.

To maintain backwards compatibility, we should make sure that requests that only try to update the language don't fail for admin users, and requests that have the `language` in the payload don't fail for offline users.

File: api/src/controllers/login.js (hunk 1 above)
Comment (dianabarsan) [reply api/src/controllers/login.js:178]: I agree. If we don't use `userSettings` anymore, we shouldn't request them. 
We should refactor `createAdmin` to deal with existent accounts and create new ones. We could just keep the one condition here: 
```
if (auth.isDbAdmin(userCtx)) {
  return users.createAdmin(userCtx);
}
```

File: api/src/controllers/login.js (hunk 2 above)
Comment (dianabarsan) [reply api/src/controllers/login.js:268]: I think this is a good solution, however the body that this endpoint receives is always empty: 
https://github.com/medic/cht-core/blob/3222eb05cb8f8248485b7168505f8cfc6e4c1c31/api/src/public/login/script.js#L55

We'd need to add a payload with the selected locale there: 
```
const payload = JSON.stringify({ locale: selectedLocale });
request('POST', url, payload, xmlhttp => {
```

File: api/tests/mocha/controllers/login.spec.js [hunk 5]
```diff
@@ -578,16 +580,12 @@ describe('login controller', () => {
...
       return controller.post(req, res).then(() => {
         chai.expect(cookie.callCount).to.equal(3);
         chai.expect(cookie.args[0][0]).to.equal('AuthSession');
         chai.expect(cookie.args[1][0]).to.equal('userCtx');
         chai.expect(cookie.args[2][0]).to.equal('locale');
         chai.expect(cookie.args[2][1]).to.equal('fr');
-        chai.expect(update.callCount).to.equal(1);
-        chai.expect(update.args[0][0]).to.equal('sharon');
-        chai.expect(update.args[0][1].language).to.equal('fr');
       });
```
Comment (dianabarsan) [inline api/tests/mocha/controllers/login.spec.js:589]: Please keep the `update.callCount` assertion here, to show that we're not actually updating user-settings.

File: api/tests/mocha/services/users.spec.js [hunk 6]
```diff
@@ -75,6 +75,12 @@ describe('Users service', () => {
       chai.expect(settings.fullname).to.equal('John');
     });
 
+    it('does not reassign language', () => {
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:78]: Please add some tests here that make sure we maintain backwards compatibility: 
1. Online user requests with just `language` field shouldn't reject with 400.
2. Offline user requests that include `language` field shouldn't reject with 401.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Good start! Thanks for picking this up! 
I'm sharing @garethbowen 's concerns about language selection when logging in with a token, and agree with solving that in a separate issue. 
I added some requests about backwards compatibility inline.

**Comment by dianabarsan** [comment]
> One question I do have is if we need to include an api migration script with these changes to remove the language property from the existing settings for each user in the medic db? I do not think this data is getting used anymore, but I am not sure if this is the kind of thing that usually gets a migration script...

This is the kind of thing that usually gets a migration script. And given that there aren't so many users (statistically), it would be a relatively cheap migration to run. 
However, we should make sure no other software is using those properties. 

Now that I think about it, from a data analysis standpoint, if we remove this field and the interaction with API, we no longer know which language the users are using? What do you think, @garethbowen . 
We should probably include the language in telemetry, but that means we'd have 1 month stale information.
Plenty of times, bugs would only surface when specific languages were used (think translation / message format bugs), without knowing what languages the users are using (with relatively up to date information), those bugs will be significantly more challenging to understand.

**Comment by garethbowen** [comment]
@dianabarsan Good point. Perhaps we should add a new issue to store the selected language code in telemetry and feedback docs?

**Comment by dianabarsan** [comment]
Added issue about storing language in telemetry and feedback docs: https://github.com/medic/cht-core/issues/7089

**Comment by jkuester** [comment]
@dianabarsan sorry for the delay on this PR! My time has been really tight for the past couple of weeks.

This code is ready to be reviewed again when you get the chance. I was not sure, from the above comments, if we needed to wait on the migration script until #7089 is addressed. I had some time to tinker with it this afternoon, and so I have included the migration script in this PR, but I can revert that commit if it needs to wait for #7089.

File: api/tests/mocha/services/users.spec.js [hunk 7]
```diff
@@ -1347,7 +1380,7 @@ describe('Users service', () => {
           chai.expect(err).to.deep.include({ code: 400 });
           chai.expect(db.medic.get.callCount).to.equal(1);
           chai.expect(db.medic.get.args[0]).to.deep.equal(['org.couchdb.user:my_user']);
-          chai.expect(db.users.get.callCount).to.equal(1);
+          chai.expect(db.users.get.callCount).to.equal(2);
           chai.expect(db.users.get.args[0]).to.deep.equal(['org.couchdb.user:my_user']);
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:1384]: Please add an assertion about the second call to `db.users.get`.

File: api/src/migrations/remove-user-language.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      if (!rowsWithLanguage.length) {
+        return fixUsers(skipSize + BATCH_SIZE);
+      }
+      return removeLanguage(rowsWithLanguage)
+        .then(() => fixUsers(skipSize + BATCH_SIZE));
+    });
+};
+
+module.exports = {
+  name: 'remove-user-language',
```
Comment (dianabarsan) [inline api/src/migrations/remove-user-language.js:33]: Thanks a lot for the migration and the integration test. 
We should also add unit tests to cover every edge case and check that batching works. Please check the other migrations unit tests here: https://github.com/medic/cht-core/tree/master/api/tests/mocha/migrations

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Great work! Thanks for the migration and the backwards compatibility fix. 
I just have a couple of suggestions inline, but nothing major!

File: api/src/migrations/remove-user-language.js (hunk 8 above)
Comment (jkuester) [reply api/src/migrations/remove-user-language.js:34]: Hmm, maybe I am missing something, but I did add some unit tests here: https://github.com/medic/cht-core/pull/7081/files#diff-7e520661b99ef648535ac94ed1cb78f3393f1d9906ae0968695c91c3b769c638

I have gone over them again and tweaked them up a bit, but I have not been able to really come up with any more test cases that seem useful.  Let me know if I have the wrong approach here or if there are other test cases that we need!

Comment (dianabarsan) [reply api/src/migrations/remove-user-language.js:34]: The one test you added is an integration test. It doesn't cover batching, it doesn't cover error handling. We should at least test that batching works correctly and that we handle errors (at least that we don't continue to query if an error happens). 

Please have a look at the folder I linked you in my comment to see how some other migrations are unit tested (for example: https://github.com/medic/cht-core/blob/master/api/tests/mocha/migrations/convert-translation-messages-fix.spec.js) 

Thanks!

Comment (jkuester) [reply api/src/migrations/remove-user-language.js:34]: Whoops!  In my comment above I linked back to the wrong file (the integration tests).  I mean to put this link: https://github.com/medic/cht-core/pull/7081/files#diff-f9e2a40a58146b6c05a82dddd7209eaa1b0ed016e29cd1dbaa6341a731e404e6

(Or here is a non-PR link: https://github.com/jkuester/cht-core/blob/master/api/tests/mocha/migrations/remove-user-language.spec.js)  I think I have it in the correct folder `api/tests/mocha/migrations` and I tried to model it after the other tests that were there.

Comment (dianabarsan) [reply api/src/migrations/remove-user-language.js:34]: Oh, this is awesome, thanks!

File: api/tests/mocha/migrations/remove-user-language.spec.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+      doc: {
+        _id: 'org.couchdb.user:noLang0',
+      }
+    };
+    const userNoLang1 = { };
+
+    userQuery.onFirstCall().resolves({ rows: [ userLang0, userNoLang0, userLang1, userNoLang1 ] });
+    userQuery.onSecondCall().resolves({ rows: [ userLang3 ] });
+    userQuery.onThirdCall().resolves({ });
+    bulkDocs.returns(Promise.resolve());
```
Comment (dianabarsan) [inline api/tests/mocha/migrations/remove-user-language.spec.js:71]:
```suggestion
    bulkDocs.resolves();
```

File: api/tests/mocha/migrations/remove-user-language.spec.js [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+    return migration.run().then(() => {
+      chai.expect(userQuery.callCount).to.equal(3);
+      chai.expect(userQuery.calledWithExactly(docByType, expectedOptions)).to.be.true;
+      expectedOptions.skip = 100;
+      chai.expect(userQuery.calledWithExactly(docByType, expectedOptions)).to.be.true;
+      expectedOptions.skip = 200;
+      chai.expect(userQuery.calledWithExactly(docByType, expectedOptions)).to.be.true;
```
Comment (dianabarsan) [inline api/tests/mocha/migrations/remove-user-language.spec.js:79]: `calledWithExactly` doesn't actually show which call had these parameters. 
For more strict assertions (preferred) we could do: 
```suggestion
      chai.expect(userQuery.args[0]).to.deep.equal([docByType, expectedOptions]);
      expectedOptions.skip = 100;
      chai.expect(userQuery.args[1]).to.deep.equal([docByType, expectedOptions]);
      expectedOptions.skip = 200;
      chai.expect(userQuery.args[2]).to.deep.equal([docByType, expectedOptions]);
```

File: api/tests/mocha/migrations/remove-user-language.spec.js [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+      expectedOptions.skip = 100;
+      chai.expect(userQuery.calledWithExactly(docByType, expectedOptions)).to.be.true;
+      chai.expect(bulkDocs.callCount).to.equal(0);
+    });
+  });
+
+  it('should throw an error if one occurs when querying', () => {
+    const message = 'Some Error';
+    userQuery.returns(Promise.reject(message));
```
Comment (dianabarsan) [inline api/tests/mocha/migrations/remove-user-language.spec.js:113]:
```suggestion
    userQuery.rejects(message);
```

**Review by dianabarsan — COMMENTED** [review]
Really like the unit test! Awesome work!

File: api/src/services/users.js [hunk 12]
```diff
@@ -518,11 +518,15 @@ module.exports = {
...
-          .then(() => createUser(data, {}))
-          .then(() => createUserSettings(data, {}));
+      .catch(err => {
+        if(err.status && err.status === 404) {
+          const data = { username: userCtx.name, roles: ['admin'] };
+          return validateNewUsername(userCtx.name)
+            .then(() => createUser(data, {}))
+            .then(() => createUserSettings(data, {}));
+        } else {
+          return Promise.reject(err);
```
Comment (dianabarsan) [inline api/src/services/users.js:528]: I _think_ your intention in the condition was to null check `err`, right? Otherwise `err.status && err.status === 404` is kind of redundant. We can also make a slight improvement and remove the unnecessary else
```suggestion
        if (err && err.status === 404) {
          const data = { username: userCtx.name, roles: ['admin'] };
          return validateNewUsername(userCtx.name)
            .then(() => createUser(data, {}))
            .then(() => createUserSettings(data, {}));
        }
        return Promise.reject(err);
```

File: api/tests/mocha/migrations/remove-user-language.spec.js [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    });
+  });
+
+  it('should throw an error if one occurs when querying', () => {
+    const message = 'Some Error';
+    userQuery.rejects(message);
+
+    return migration.run()
+      .catch((error) => {
+        chai.expect(userQuery.callCount).to.equal(1);
```
Comment (dianabarsan) [inline api/tests/mocha/migrations/remove-user-language.spec.js:117]: I'm super sorry I didn't see this before. 
We need to add an additional step here, to make sure that the code is actually throwing an error. If `migration.run()` doesn't throw, we never execute the code in the catch and never run any assertions, and the test will still pass. 
```suggestion
    return migration
      .run()
      .then(() => chai.assert.fail('should have thrown'))
      .catch((error) => {
        chai.expect(userQuery.callCount).to.equal(1);
```

This also applies to the other test that checks that errors are thrown.

File: api/tests/mocha/migrations/remove-user-language.spec.js [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        chai.expect(userQuery.calledWithExactly(docByType, expectedOptions)).to.be.true;
```
Comment (dianabarsan) [inline api/tests/mocha/migrations/remove-user-language.spec.js:118]: Please use `.args[..]` and `.to.deep.equal` everywhere instead of `calledWithExactly`.

File: api/src/migrations/remove-user-language.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
+const db = require('../db');
+const BATCH_SIZE = 100;
+
+const removeLanguage = (docs) => {
+  docs.forEach(doc => delete doc.language);
```
Comment (dianabarsan) [inline api/src/migrations/remove-user-language.js:5]: I just realized we have another regression because of this, in webapp: 

image

The language dropdown used to get the value from the user settings document: https://github.com/medic/cht-core/blob/master/webapp/src/ts/modals/edit-user/edit-user.component.ts#L34 . Now it's always blank (duh). 
We should change this to get the language from the cookie.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Great work! 
I found one minor regression (that I didn't notice before, sorry!) and I have a couple of suggestions in the test cases. 

Other than that, please rebase off of current master to fix your build. 

Thanks!

File: api/tests/mocha/migrations/remove-user-language.spec.js (hunk 13 above)
Comment (jkuester) [reply api/tests/mocha/migrations/remove-user-language.spec.js:117]: Whoops!  Good call. I have gone ahead and updated similar places in the other migration tests too.

File: api/src/migrations/remove-user-language.js (hunk 15 above)
Comment (jkuester) [reply api/src/migrations/remove-user-language.js:5]: Okay, I am pretty sure I have got this fixed correctly (full disclosure: I never touched Angular before last week...).  I also just completely removed the call to the user settings from the LanguageService since it seemed pointless now.

File: api/src/services/users.js [hunk 16]
```diff
@@ -48,11 +48,14 @@ const SETTINGS_EDITABLE_FIELDS = RESTRICTED_SETTINGS_EDITABLE_FIELDS.concat([
...
 
+// No longer used, but allowed for backwards compatibility
+const LEGACY_FIELDS = ['language'];
+
 const ALLOWED_RESTRICTED_EDITABLE_FIELDS =
   RESTRICTED_SETTINGS_EDITABLE_FIELDS.concat(RESTRICTED_USER_EDITABLE_FIELDS, META_FIELDS);
 
 const illegalDataModificationAttempts = data =>
-  Object.keys(data).filter(k => !ALLOWED_RESTRICTED_EDITABLE_FIELDS.includes(k));
+  Object.keys(data).filter(k => !ALLOWED_RESTRICTED_EDITABLE_FIELDS.concat(LEGACY_FIELDS).includes(k));
```
Comment (dianabarsan) [inline api/src/services/users.js:58]: I think a change got lost in the merge here, removing `language` from `RESTRICTED_SETTINGS_EDITABLE_FIELDS`. Right now, you end up editing the language after all.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Great work on updating webapp! 
I think I found a merge problem, once that's fixed, I think this is good to go!

File: api/tests/mocha/migrations/extract-user-settings.spec.js [hunk 17]
```diff
@@ -75,11 +78,14 @@ describe('extract-user-settings migration', () => {
     const list = sinon.stub(db.users, 'allDocs').callsArgWith(1, null, { rows: [ ddoc, userA, userB ] });
     sinon.stub(db.medic, 'get').callsArgWith(1, { error: 'not_found'});
     sinon.stub(db.medic, 'put').callsArgWith(1, 'boom');
-    migration.run().catch(err => {
-      chai.expect(err).to.equal('boom');
-      chai.expect(list.callCount).to.equal(1);
-      done();
-    });
+    migration
```
[the remark this answers is not in the document]
Comment (jkuester) [reply api/tests/mocha/migrations/extract-user-settings.spec.js:81]: Shoot, missed that.  Thanks for the link too, that is good to know!  I have gone ahead and reverted the changes to all the tests that were calling `done()`

File: api/src/services/users.js (hunk 16 above)
Comment (jkuester) [reply api/src/services/users.js:57]: Wow.  I really wish I had done a merge instead of a rebase now.  (Would have made it easier to figure out what happened here. I don't remember any conflicts while rebasing, but apparently something got messed up...)  Anyway, I have removed 'language' again from the RESTRICTED_SETTINGS_EDITABLE_FIELDS and have read back through all the other changes looking for anything else that got dropped (but did not find anything).

Comment (dianabarsan) [reply api/src/services/users.js:57]: Yea :) 
A unit test should have caught this.

File: api/tests/mocha/services/users.spec.js [hunk 18]
```diff
@@ -1288,6 +1288,33 @@ describe('Users service', () => {
...
+    it('succeeds if only language is defined by a user with full access', () => {
+      const data = {
+        language: 'es'
+      };
+      service.__set__('validateUser', sinon.stub().resolves({}));
+      service.__set__('validateUserSettings', sinon.stub().resolves({}));
+      sinon.stub(db.medic, 'put').resolves({});
+      sinon.stub(db.users, 'put').resolves({});
+      return service.updateUser('paul', data, true).then(() => {
+        chai.expect(db.medic.put.callCount).to.equal(1);
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:1300]: Re: the merge problem. 
I we had had assertions here about how the arguments to put looked like, and making sure it doesn't contain `language`, the unit test would have caught the merge issue.

Comment (jkuester) [reply api/tests/mocha/services/users.spec.js]: Good call. I have updated these tests to assert the arguments passed to medic/users.

**Review by dianabarsan — APPROVED** [review]
An excellent first contribution!
